tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,253 @@
+		Log:              logger,
+		messageReceiver:  messageReceiver,
+		messageSender:    messageSender,
+		beaconCalculator: beaconCalculator,
+		atxDB:            atxDB,
+		layerTicker:      layerTicker,
+		networkDelta:     time.Duration(conf.WakeupDelta) * time.Second,
+		currentRounds:    make(map[types.EpochID]int),
+		timelyProposals:  make(map[EpochRoundPair][]types.ATXID),
+		delayedProposals: make(map[EpochRoundPair][]types.ATXID),
+		lateProposals:    make(map[EpochRoundPair][]types.ATXID),
+	}
+}
+
+// Start starts listening for layers and outputs.
+func (tb *TortoiseBeacon) Start() error {
+	tb.Log.Info("Starting %v", protoName)
+
+	go tb.tickLoop()
+	go func() {
+		if err := tb.listen(); err != nil {
+			// TODO: handle error
+			return
+		}
+	}()
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) Get(epochID types.EpochID) types.Hash32 {
+	tb.beaconsMu.RLock()
+	ch := tb.beacons[epochID]
+	tb.beaconsMu.RUnlock()
+
+	beacon := <-ch
+	return beacon
+}
+
+func (tb *TortoiseBeacon) listen() error {
+	for {
+		m := tb.messageReceiver.Receive()
+
+		if err := tb.handleMessage(m); err != nil {
+			return err
+		}
+	}
+}
+
+// listens to new layers.
+func (tb *TortoiseBeacon) tickLoop() {
+	for {
+		select {
+		case layer := <-tb.layerTicker:
+			go tb.onTick(layer)
+		case <-tb.CloseChannel():
+			return
+		}
+	}
+}
+
+// the logic that happens when a new layer arrives.
+// this function triggers the start of new CPs.
+func (tb *TortoiseBeacon) onTick(layer types.LayerID) {
+	tb.layerMu.Lock()
+	if layer > tb.lastLayer {
+		tb.lastLayer = layer
+	}
+
+	tb.layerMu.Unlock()
+
+	epoch := layer.GetEpoch()
+	tb.Debug("tortoise beacon got tick at layer %v epoch %v, sleeping for %v", layer, epoch, tb.networkDelta)
+
+	if layer.GetEpoch().IsGenesis() {
+		tb.With().Info("not starting tortoise beacon since we are in genesis epoch", layer)
+		return
+	}
+
+	tb.beaconsMu.Lock()
+
+	if _, ok := tb.beacons[epoch]; ok {
+		tb.beaconsMu.Unlock()
+
+		// Already handling this epoch
+		return
+	}
+
+	tb.beacons[epoch] = make(chan types.Hash32, 1)
+	tb.beaconsMu.Unlock()
+
+	go func() {
+		tb.roundTicker(epoch)
+
+		tb.timelyProposalsMu.Lock()
+		beacon := tb.beaconCalculator.CalculateBeacon(tb.timelyProposals)
+		tb.timelyProposalsMu.Unlock()
+
+		tb.beaconsMu.Lock()
+		tb.beacons[epoch] <- beacon
+		close(tb.beacons[epoch])
+		tb.beaconsMu.Unlock()
+	}()
+
+	atxList := tb.atxDB.GetEpochAtxs(epoch - 1)
+	m := NewMessage(epoch, 0, atxList)
+	tb.messageSender.Send(m)
+}
+
+func (tb *TortoiseBeacon) handleMessage(m Message) error {
+	epoch := m.Epoch()
+	tb.classifyMessage(m, epoch)
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) classifyMessage(m Message, epoch types.EpochID) {
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	pair := EpochRoundPair{
+		EpochID: epoch,
+		Round:   currentRound,
+	}
+
+	if m.Round() <= currentRound-1 {
+		tb.timelyProposalsMu.Lock()
+		tb.timelyProposals[pair] = m.Payload()
```

Comment:
in this implementation the `timelyProposals` is used to construct the next message and therefore it should be a union of all timely proposals